pyanaconda/modules/payload/payload.py

```diff
@@ -42,43 +43,108 @@ def publish(self):
         DBus.publish_object(PAYLOAD.object_path, PayloadInterface(self))
         DBus.register_service(PAYLOAD.service_name)
 
-    def _publish_handler(self):
-        """Publish handler as soon as we know which one to chose"""
-        self._payload_handler.publish()
-
     @property
     def kickstart_specification(self):
         """Return the kickstart specification."""
         return PayloadKickstartSpecification
 
+    @property
+    def payload_handler(self):
+        """Get payload handler.
+
+        Handlers are handling the installation process.
+
+        There are a few types of handler e.g.: DNF, LiveImage...
+        """
+        if self._payload_handler is None:
+            raise HandlerNotSetError()
+        else:
+            return self._payload_handler
+
+    def set_payload_handler(self, payload_handler):
+        """Set payload handler."""
+        self._payload_handler = payload_handler
+        log.debug("Payload handler %s used.", payload_handler.__class__.__name__)
+
+    def get_active_handler_path(self):
+        """Get path of the active payload handler.
+
+        :rtype: str
+        """
+        try:
+            return self.payload_handler.get_handler_path()
+        except HandlerNotSetError:
+            return ""
+
     def process_kickstart(self, data):
         """Process the kickstart data."""
         log.debug("Processing kickstart data...")
 
-        self._create_correct_handler(data)
-        self._publish_handler()
+        handler_class = self._get_correct_handler_class(data)
+
+        if not handler_class:
+            log.warning("No handler was created. Kickstart data passed in are lost.")
+            return
 
-        self._payload_handler.process_kickstart(data)
+        self._initialize_handler(handler_class)
 
-    def _create_correct_handler(self, data):
+        self.payload_handler.process_kickstart(data)
+
+    def _get_correct_handler_class(self, data):
         if data.liveimg.seen:
-            self._payload_handler = LiveImageHandlerModule()
+            return LiveImageHandlerModule
         else:
-            self._payload_handler = DNFHandlerModule()
+            return None
+
+    def _initialize_handler(self, handler_class):
```

Comment:
I would prefer to create handlers based on a requested type.

Reply:
Fixed.